apps/builder/app/canvas/collapsed.ts

```diff
@@ -0,0 +1,122 @@
+import { collapsedAttribute } from "@webstudio-is/react-sdk";
+
+const instanceIdSet = new Set<string>();
+let rafHandle: number;
+
+const recalculate = () => {
+  const instanceIds = Array.from(instanceIdSet);
+  instanceIdSet.clear();
+  if (instanceIds.length === 0) {
+    return;
+  }
+
+  /**
+   *  Selector to find element itself or common ancestor
+   *  (finding common ancestor is suboptimal solution but it almost never happens)
+   **/
+  const eltSelector =
+    instanceIds.length === 1
+      ? `[data-ws-id=${instanceIds[0]}]`
+      : `[data-ws-id]${instanceIds
+          .map((instanceId) => `:has([data-ws-id=${instanceId}])`)
+          .join("")}`;
+
+  const elements: Element[] = [];
+
+  // Element itself or common ancestor or body
+  const baseElement =
+    Array.from(document.querySelectorAll(eltSelector)).pop() ?? document.body;
+  elements.push(baseElement);
+
+  const descendants = baseElement.querySelectorAll("[data-ws-id]");
+  elements.push(...descendants);
+
+  const elementsToRecalculate: HTMLElement[] = [];
+  const parentsWithAbsoluteChildren = new Map<HTMLElement, number>();
+
+  for (const element of elements) {
+    if (!(element instanceof HTMLElement)) {
+      continue;
+    }
+
+    if (element.tagName === "_IMG") {
```

Comment:
I think we should ignore all void elements https://developer.mozilla.org/en-US/docs/Glossary/Void_element